File: crates/compiler/zo-benches/benches/spectralnorm/spectralnorm.c
// spectralnorm (benchmarksgame): largest eigenvalue of the infinite
// matrix A[i][j] = 1/((i+j)(i+j+1)/2 + i + 1), via 10 power-method
// iterations of AᵀA on the all-ones vector. Serial port (the provided
// OpenMP/SSE version uses Linux-only sched.h/cpu_set_t). Prints
// sqrt((u·v)/(v·v)). Reads n from argv (default 100).

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

static double a(int i, int j) {
  return 1.0 / (((i + j) * (i + j + 1) / 2) + i + 1);
}

static void mult_av(const double *v, double *out, int n) {
  for (int i = 0; i < n; i++) {
    double sum = 0.0;
    for (int j = 0; j < n; j++)
      sum += a(i, j) * v[j];
    out[i] = sum;
  }
}

static void mult_atv(const double *v, double *out, int n) {
  for (int i = 0; i < n; i++) {
    double sum = 0.0;
    for (int j = 0; j < n; j++)
      sum += a(j, i) * v[j];
    out[i] = sum;
  }
}

static void mult_atav(const double *v, double *out, double *tmp, int n) {
  mult_av(v, tmp, n);
  mult_atv(tmp, out, n);
}

int main(int argc, char **argv) {
  const int n = argc > 1 ? atoi(argv[1]) : 100;

  double *u = malloc((size_t)n * sizeof(double));
  double *v = malloc((size_t)n * sizeof(double));
  double *tmp = malloc((size_t)n * sizeof(double));

  for (int i = 0; i < n; i++) {
    u[i] = 1.0;
    v[i] = 0.0;
  }

  for (int it = 0; it < 10; it++) {
    mult_atav(u, v, tmp, n);
    mult_atav(v, u, tmp, n);
  }

  double v_bv = 0.0, vv = 0.0;
  for (int i = 0; i < n; i++) {
    v_bv += u[i] * v[i];
    vv += v[i] * v[i];
  }

  printf("%.9f\n", sqrt(v_bv / vv));

  free(u);
  free(v);
  free(tmp);
  return 0;
}

File: crates/compiler/zo-benches/benches/spectralnorm/spectralnorm.go
// spectralnorm (benchmarksgame): largest eigenvalue of A[i][j] =
// 1/((i+j)(i+j+1)/2 + i + 1), via 10 power-method iterations of AᵀA
// on the all-ones vector. Serial port. Reads n from argv (default 100).

package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
)

func a(i, j int) float64 {
	return 1.0 / float64((i+j)*(i+j+1)/2+i+1)
}

func multAv(v, out []float64) {
	for i := range out {
		sum := 0.0
		for j, vj := range v {
			sum += a(i, j) * vj
		}
		out[i] = sum
	}
}

func multAtv(v, out []float64) {
	for i := range out {
		sum := 0.0
		for j, vj := range v {
			sum += a(j, i) * vj
		}
		out[i] = sum
	}
}

func multAtAv(v, out, tmp []float64) {
	multAv(v, tmp)
	multAtv(tmp, out)
}

func main() {
	n := 100
	if len(os.Args) > 1 {
		if x, err := strconv.Atoi(os.Args[1]); err == nil {
			n = x
		}
	}

	u := make([]float64, n)
	v := make([]float64, n)
	tmp := make([]float64, n)
	for i := range u {
		u[i] = 1.0
	}

	for i := 0; i < 10; i++ {
		multAtAv(u, v, tmp)
		multAtAv(v, u, tmp)
	}

	vBv, vv := 0.0, 0.0
	for i := 0; i < n; i++ {
		vBv += u[i] * v[i]
		vv += v[i] * v[i]
	}

	fmt.Printf("%0.9f\n", math.Sqrt(vBv/vv))
}
